File: backend/compfest-ai/app/services/pinecone_service.py
from pinecone.grpc import PineconeGRPC as Pinecone
from pinecone import ServerlessSpec
from flask import current_app, jsonify, Blueprint, request
from enum import Enum

class ManageIndexEnum(Enum):
    CREATE = "create"
    DELETE = "delete"

def initialize_pinecone():
    return Pinecone(
        api_key=current_app.config['PINECONE_API_KEY']
    )

def manage_index(mode: str):
    pc = initialize_pinecone()
    
    if mode == ManageIndexEnum.CREATE.value:
        index_name = current_app.config['PINECONE_INDEX_NAME']
        if not index_name:
            return "Index name not provided", 400
        if index_name in pc.list_indexes().names():
            return "Index already exists", 400

        pc.create_index(
            name=index_name,
            dimension=current_app.config['PINECONE_DIMENSIONS'],
            metric=current_app.config['PINECONE_SIMILARITY_METRIC'],
            spec=ServerlessSpec(
                cloud=current_app.config['PINECONE_CLOUD_PROVIDER'],
                region=current_app.config['PINECONE_REGION']
            )
        )
        return "Index created successfully", 201

    elif mode == ManageIndexEnum.DELETE.value:
        index_name = current_app.config['PINECONE_INDEX_NAME']
        if not index_name:
            return "Index name not provided", 400
        if index_name not in pc.list_indexes().names:
            return "Index does not exist", 400

        pc.delete_index(index_name)
        return "Index deleted successfully", 200

    else:
        return "Invalid action", 400

File: frontend/compfest-ai/src/lib/service/pinecone.service.ts
// import { Pinecone } from "@pinecone-database/pinecone";
// import type { ManagePineconeIndexType } from "@lib/config";
// import { ManagePineconeIndex, PineconeNamespaces } from "@lib/config";
// import { PINECONE_CONFIG } from "@lib/config";
// import dotenv from "dotenv";
//
// dotenv.config();
//
// export class PineconeService {
//   private static instance: PineconeService;
//
//   private connection: Pinecone;
//
//   constructor() {
//     this.connection = new Pinecone({
//       apiKey: process.env.PINECONE_API_KEY as string,
//     });
//
//     if (!this.connection) {
//       throw new Error("Failed to initialize Pinecone connection");
//     }
//   }
//
//   public static getInstance(): PineconeService {
//     if (!PineconeService.instance) {
//       PineconeService.instance = new PineconeService();
//     }
//     return PineconeService.instance;
//   }
//
//   async queryEmbeddings({
//     queryText,
//     namespace,
//     topK = PINECONE_CONFIG.similarityQuery.topK,
//   }: {
//     topK?: number;
//     queryText: string;
//     namespace: string;
//   }): Promise<any> {
//     throw new Error("Not yet implemented");
//     // const queryEmbedding = await openAiService.createEmbeddings({ queryText });
//
//     // const queryResult = await this.connection
//     //   .index(PINECONE_CONFIG.indexName)
//     //   .namespace(namespace)
//     //   .query({
//     //     ...PINECONE_CONFIG.similarityQuery,
//     //     topK: topK,
//     //     vector: queryEmbedding.data[0].embedding,
//     //   });
//     //
//     // return {
//     //   queryText,
//     //   queryResult,
//     // };
//   }
//
//   async queryUserById({ userId }: { userId: string }) {
//     throw new Error("Not updated");
//
//     const index = this.connection.Index(PINECONE_CONFIG.indexName);
//     try {
//       const queryResponse = await index
//         .namespace(PineconeNamespaces.user)
//         .fetch([userId]);
//
//       if (!queryResponse.records) {
//         throw new Error("User not found");
//       }
//
//       return queryResponse.records;
//     } catch (error) {
//       throw new Error(`Error querying user by ID: ${userId}. Error: ${error}`);
//     }
//   }
//
//   async manageIndex(action: ManagePineconeIndexType): Promise<string> {
//     try {
//       const indextExists = (await this.connection.listIndexes()).indexes?.some(
//         (index) => index.name === PINECONE_CONFIG.indexName,
//       );
//
//       if (action == ManagePineconeIndex.create) {
//         if (indextExists) {
//           throw new Error("Index already exists");
//         }
//
//         await this.connection.createIndex({
//           name: PINECONE_CONFIG.indexName,
//           dimension: PINECONE_CONFIG.dimensions,
//           metric: PINECONE_CONFIG.metric,
//           spec: {
//             serverless: {
//               cloud: PINECONE_CONFIG.cloud,
//               region: PINECONE_CONFIG.region,
//             },
//           },
//         });
//         return "Success";
//       }
//
//       if (action == ManagePineconeIndex.delete) {
//         if (!indextExists) {
//           throw new Error("Index does not exist");
//         }
//         await this.connection.deleteIndex(PINECONE_CONFIG.indexName);
//         return "Success";
//       }
//
//       throw new Error("Invalid action");
//     } catch (error) {
//       throw error;
//     }
//   }
// }
//
// export const pineconeService = PineconeService.getInstance();
